smart-contracts/contracts/mixins/MixinTransfer.sol

```diff
@@ -0,0 +1,183 @@
+    address _recipient,
+    uint _tokenId
+  )
+    public
+    payable
+    onlyIfAlive
+    hasValidKey(_from)
+    onlyKeyOwnerOrApproved(_tokenId)
+  {
+    require(_recipient != address(0), 'INVALID_ADDRESS');
+
+    Key storage fromKey = _getKeyFor(_from);
+    Key storage toKey = _getKeyFor(_recipient);
+
+    uint previousExpiration = toKey.expirationTimestamp;
+
+    if (previousExpiration == 0) {
+      // The recipient did not have a key previously
+      _addNewOwner(_recipient);
+      _setKeyOwner(_tokenId, _recipient);
+      toKey.tokenId = _tokenId;
+    }
+
+    if (previousExpiration <= block.timestamp) {
+      // The recipient did not have a key, or had a key but it expired. The new expiration is the
+      // sender's key expiration
+      toKey.expirationTimestamp = fromKey.expirationTimestamp;
+    } else {
+      // The recipient has a non expired key. We just add them the corresponding remaining time
+      // SafeSub is not required since the if confirms `previousExpiration - block.timestamp` cannot underflow
+      toKey.expirationTimestamp = fromKey
+        .expirationTimestamp.add(previousExpiration - block.timestamp);
+    }
+    // Overwite data in all cases
+    toKey.data = fromKey.data;
+
+    // Effectively expiring the key for the previous owner
+    fromKey.expirationTimestamp = block.timestamp;
+
+    // Clear any previous approvals
+    _clearApproval(_tokenId);
+
+    // trigger event
+    emit Transfer(
+      _from,
+      _recipient,
+      _tokenId
+    );
+  }
+
+  /**
+  * @notice Transfers the ownership of an NFT from one address to another address
+  * @dev This works identically to the other function with an extra data parameter,
+  *  except this function just sets data to ''
+  * @param _from The current owner of the NFT
+  * @param _to The new owner
+  * @param _tokenId The NFT to transfer
+  */
+  function safeTransferFrom(
+    address _from,
+    address _to,
+    uint _tokenId
+  )
+    external
+    payable
+  {
+    safeTransferFrom(_from, _to, _tokenId, '');
+  }
+
+  /**
+  * @notice Transfers the ownership of an NFT from one address to another address.
+  * When transfer is complete, this functions
+  *  checks if `_to` is a smart contract (code size > 0). If so, it calls
+  *  `onERC721Received` on `_to` and throws if the return value is not
+  *  `bytes4(keccak256('onERC721Received(address,address,uint,bytes)'))`.
+  * @param _from The current owner of the NFT
+  * @param _to The new owner
+  * @param _tokenId The NFT to transfer
+  * @param _data Additional data with no specified format, sent in call to `_to`
+  */
+  function safeTransferFrom(
+    address _from,
+    address _to,
+    uint _tokenId,
+    bytes _data
+  )
+    public
+    payable
+    onlyIfAlive
+    onlyKeyOwnerOrApproved(_tokenId)
+    hasValidKey(ownerOf(_tokenId))
+  {
+    transferFrom(_from, _to, _tokenId);
+    require(_checkOnERC721Received(_from, _to, _tokenId, _data), 'NO_FALLBACK');
```

Comment:
I think we should use a new error code for this use case.

Reply:
@hardlydifficult My first step with this was to create a new error code in `UnlockErrors.sol`...however, when calling this function when  the `_to` address is a non-compliant contract, by definition that contract does not implement the `IERC721Receiver.sol` interface. So, the fallback is triggered and we receive the `'NO_FALLBACK'` error code. So, I had to modify the  `shouldFail`test to expect this error code instead. Maybe I'm missing something here though...

Reply:
The no_fallback only applies when sending `_to` our own contract.  More often it will be about transferring to a different contract.  I think the require line here should use a custom error code -- even though in your test case you still see `NO_FALLBACK`, that's unique to our contract.  More often than not we will get the default false returned instead.

Reply:
I see what you mean. Thanks, I'll make the changes now.